2-0-jdbc-api/2-1-1-product-dao/src/test/java/com/bobocode/ProductDaoTest.java

```diff
@@ -0,0 +1,453 @@
+                    "\n");
+        }
+    }
+
+    private Product generateTestProduct() {
+        return Product.builder()
+                .name(RandomStringUtils.randomAlphabetic(10))
+                .producer(RandomStringUtils.randomAlphabetic(20))
+                .price(BigDecimal.valueOf(RandomUtils.nextInt(10, 100)))
+                .expirationDate(LocalDate.ofYearDay(LocalDate.now().getYear() + RandomUtils.nextInt(1, 5),
+                        RandomUtils.nextInt(1, 365)))
+                .build();
+    }
+
+    @Test
+    @Order(1)
+    @DisplayName("Save a product")
+    void save() {
+
+        Product fanta = createTestFantaProduct();
+        int productsCountBeforeInsert = findAllFromDataBase().size();
+        productDao.save(fanta);
+        List<Product> products = findAllFromDataBase();
+
+        assertNotNull(fanta.getId());
+        assertThat(productsCountBeforeInsert + 1).isEqualTo(products.size());
+        assertTrue(products.contains(fanta));
+    }
+
+    private List<Product> findAllFromDataBase() {
+        try (Connection connection = dataSource.getConnection()) {
+            return findAllProducts(connection);
+        } catch (SQLException e) {
+            throw new DaoOperationException("Error finding all products", e);
+        }
+    }
+
+    private List<Product> findAllProducts(Connection connection) throws SQLException {
+        Statement statement = connection.createStatement();
+        ResultSet resultSet = statement.executeQuery("SELECT * FROM products;");
+        return collectToList(resultSet);
+    }
+
+    private List<Product> collectToList(ResultSet resultSet) throws SQLException {
+        List<Product> products = new ArrayList<>();
+        while (resultSet.next()) {
+            Product product = parseRow(resultSet);
+            products.add(product);
+        }
+        return products;
+    }
+
+    private Product parseRow(ResultSet resultSet) {
+        try {
+            return createFromResultSet(resultSet);
+        } catch (SQLException e) {
+            throw new DaoOperationException("Cannot parse row to create product instance", e);
+        }
+    }
+
+    private Product createFromResultSet(ResultSet resultSet) throws SQLException {
+        Product product = new Product();
+        product.setId(resultSet.getLong("id"));
+        product.setName(resultSet.getString("name"));
+        product.setProducer(resultSet.getString("producer"));
+        product.setPrice(resultSet.getBigDecimal("price").stripTrailingZeros());
+        product.setExpirationDate(resultSet.getDate("expiration_date").toLocalDate());
+        product.setCreationTime(resultSet.getTimestamp("creation_time").toLocalDateTime());
+        return product;
+    }
+
+    @Test
+    @Order(2)
+    @DisplayName("save() throws an exception when product ID is invalid")
+    void saveInvalidProduct() {
+        Product invalidTestProduct = createInvalidTestProduct();
+
+        try {
```

Comment:
@mathodcoast instead of `try` let's use AssertJ `assertThatThrownBy` or something similar. Where you can also check the exception class and message.